core/src/avm2/value.rs

```diff
@@ -183,60 +203,462 @@ pub fn abc_default_value<'gc>(
+            }
+            (_, Value::Bool(_)) => {
+                let number_other = Value::from(other.coerce_to_number(activation)?);
+
+                self.abstract_eq(&number_other, activation)
+            }
+            (Value::String(_), Value::Object(_))
+            | (Value::Number(_), Value::Object(_))
+            | (Value::Unsigned(_), Value::Object(_))
+            | (Value::Integer(_), Value::Object(_)) => {
+                //TODO: Should this be `Hint::Number`, `Hint::String`, or no-hint?
+                let primitive_other = other.coerce_to_primitive(Some(Hint::Number), activation)?;
+
+                self.abstract_eq(&primitive_other, activation)
+            }
+            (Value::Object(_), Value::String(_))
+            | (Value::Object(_), Value::Number(_))
+            | (Value::Object(_), Value::Unsigned(_))
+            | (Value::Object(_), Value::Integer(_)) => {
+                //TODO: Should this be `Hint::Number`, `Hint::String`, or no-hint?
+                let primitive_self = self.coerce_to_primitive(Some(Hint::Number), activation)?;
+
+                primitive_self.abstract_eq(other, activation)
+            }
+            _ => Ok(false),
         }
     }
 
-    pub fn as_namespace(&self) -> Result<&Namespace<'gc>, Error> {
-        match self {
-            Value::Namespace(ns) => Ok(ns),
-            _ => Err(format!("Expected Namespace, found {:?}", self).into()),
+    /// Determine if this value is abstractly less than the other.
+    ///
+    /// This abstract relational comparison algorithm is intended to match
+    /// ECMA-262 3rd edition, section 11.8.5. It returns `true`, `false`, *or*
+    /// `undefined` (to signal NaN), the latter of which we represent as `None`.
+    #[allow(clippy::float_cmp)]
+    pub fn abstract_lt(
+        &self,
+        other: &Value<'gc>,
+        activation: &mut Activation<'_, 'gc, '_>,
+    ) -> Result<Option<bool>, Error> {
+        let prim_self = self.coerce_to_primitive(Some(Hint::Number), activation)?;
+        let prim_other = other.coerce_to_primitive(Some(Hint::Number), activation)?;
+
+        if let (Value::String(s), Value::String(o)) = (&prim_self, &prim_other) {
+            return Ok(Some(s.to_string().bytes().lt(o.to_string().bytes())));
+        }
+
+        let num_self = prim_self.coerce_to_number(activation)?;
+        let num_other = prim_other.coerce_to_number(activation)?;
+
+        if num_self.is_nan() || num_other.is_nan() {
+            return Ok(None);
         }
+
+        if num_self == num_other
+            || num_self.is_infinite() && num_self.is_sign_positive()
+            || num_other.is_infinite() && num_other.is_sign_negative()
+        {
+            return Ok(Some(false));
+        }
+
+        if num_self.is_infinite() && num_self.is_sign_negative()
+            || num_other.is_infinite() && num_other.is_sign_positive()
+        {
+            return Ok(Some(true));
+        }
```

Comment:
I don't think you need these conditions either, these are all taken care of by the `num_self < num_other` below:

`f64::INFINITY < n` is false for any n
`n < f64::NEG_INFINITY` is false for any n
`f64::NEG_INFINITY < n` is true unless n is also negative infinity
`n < f64::INFINITY` is true unless n is also infinity